sql/catalog/lease: TestTxnObeysTableModificationTime failed [liveness failure caused jobs failure]

sql/catalog/lease.TestTxnObeysTableModificationTime [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10969631?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10969631?buildTab=artifacts#/) on release-23.1 @ [207e4959d30012b2887f31f32d3fdaea630496b9](https://github.com/cockroachdb/cockroach/commits/207e4959d30012b2887f31f32d3fdaea630496b9):


```
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:927
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:902
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1686
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1621
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.persistProgress
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:281
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.deleteDataAndWaitForGC
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:333
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:314
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1628
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1629
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:474
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func2
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:388
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | [...repeated from below...]
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +Wraps: (6) failed to read query result
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +Wraps: (7) candidate pg code: 57014
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +Wraps: (8) attached stack trace
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  -- stack trace:
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | github.com/cockroachdb/cockroach/pkg/util/cancelchecker.init
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	github.com/cockroachdb/cockroach/pkg/util/cancelchecker/cancel_checker.go:80
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.doInit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:6349
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.doInit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:6326
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.doInit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:6326
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.doInit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:6326
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.doInit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:6326
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.doInit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:6326
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.main
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/proc.go:233
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | runtime.goexit
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +  | 	GOROOT/src/runtime/asm_amd64.s:1594
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +Wraps: (9) query execution canceled
W230719 09:54:18.444261 93470 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=883720383116476417] 5433 +Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.withPrefix (7) *pgerror.withCandidateCode (8) *withstack.withStack (9) *errutil.leafError
I230719 09:54:18.446704 81233 server/server_controller_orchestration.go:263  [T1,n1] 5434  server controller shutting down ungracefully
I230719 09:54:18.446745 81233 server/server_controller_orchestration.go:274  [T1,n1] 5435  waiting for tenant servers to report stopped
W230719 09:54:18.446814 81233 server/server_sql.go:1704  [T1,n1] 5436  server shutdown without a prior graceful drain
--- FAIL: TestTxnObeysTableModificationTime (1.79s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnObeysTableModificationTime.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29922

sql/catalog/lease.TestTxnObeysTableModificationTime [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11107724?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11107724?buildTab=artifacts#/) on release-23.1 @ [4a64824134c25d6623e15ad8b3bf7627a26a65d5](https://github.com/cockroachdb/cockroach/commits/4a64824134c25d6623e15ad8b3bf7627a26a65d5):


```
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:927
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:902
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1686
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1621
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.persistProgress
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:281
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.deleteDataAndWaitForGC
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:333
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:314
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1628
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1629
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:474
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func2
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:388
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +Wraps: (6) failed to read query result
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +Wraps: (7) candidate pg code: 57014
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +Wraps: (8) attached stack trace
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  -- stack trace:
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | github.com/cockroachdb/cockroach/pkg/util/cancelchecker.init
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	github.com/cockroachdb/cockroach/pkg/util/cancelchecker/cancel_checker.go:80
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.doInit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:6349
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.doInit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:6326
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.doInit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:6326
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.doInit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:6326
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.doInit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:6326
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.doInit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:6326
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.main
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/proc.go:233
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | runtime.goexit
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +  | 	GOROOT/src/runtime/asm_arm64.s:1172
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +Wraps: (9) query execution canceled
W230728 23:57:53.543753 114084 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886434278224691201] 5596 +Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.withPrefix (7) *pgerror.withCandidateCode (8) *withstack.withStack (9) *errutil.leafError
I230728 23:57:53.547921 93516 server/server_controller_orchestration.go:268  [T1,n1] 5597  server controller shutting down ungracefully
I230728 23:57:53.547961 93516 server/server_controller_orchestration.go:279  [T1,n1] 5598  waiting for tenant servers to report stopped
W230728 23:57:53.548017 93516 server/server_sql.go:1704  [T1,n1] 5599  server shutdown without a prior graceful drain
--- FAIL: TestTxnObeysTableModificationTime (3.74s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #107736 sql/catalog/lease: TestTxnObeysTableModificationTime failed [C-test-failure O-robot T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnObeysTableModificationTime.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/catalog/lease.TestTxnObeysTableModificationTime [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11111870?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11111870?buildTab=artifacts#/) on release-23.1 @ [8b970683af4630cbb30b4c92f0be86cfadd97993](https://github.com/cockroachdb/cockroach/commits/8b970683af4630cbb30b4c92f0be86cfadd97993):


```
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1698
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:965
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/kv/txn.go:928
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/kv.runTxn
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:964
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:927
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:902
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1686
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1621
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.persistProgress
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:281
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.deleteDataAndWaitForGC
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:333
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:314
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1628
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1629
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:474
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func2
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:388
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | runtime.goexit
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +  | 	GOROOT/src/runtime/asm_amd64.s:1594
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +Wraps: (2) job 886590545141465089
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +Wraps: (3) context canceled
W230729 13:12:42.355951 99564 sql/gcjob/gc_job_utils.go:299  [T1,n1,job=SCHEMA CHANGE GC id=886590545141465089] 5497 +Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errors.errorString
I230729 13:12:42.356293 87730 server/start_listen.go:103  [T1,n1] 5499  server shutting down: instructing cmux to stop accepting
I230729 13:12:42.356377 88770 sql/stats/automatic_stats.go:612  [T1,n1] 5500  quiescing stats garbage collector
I230729 13:12:42.356687 88142 rpc/context.go:2318  [T1,n1,rnode=1,raddr=127.0.0.1:40433,class=default,rpc] 5501  connection heartbeat loop ended with err: <nil>
W230729 13:12:42.356933 88464 sql/sqlliveness/slinstance/slinstance.go:334  [T1,n1] 5502  exiting heartbeat loop
W230729 13:12:42.356959 88464 sql/sqlliveness/slinstance/slinstance.go:321  [T1,n1] 5503  exiting heartbeat loop with error: node unavailable; try another peer
E230729 13:12:42.356982 88464 server/server_sql.go:509  [T1,n1] 5504  failed to run update of instance with new session ID: node unavailable; try another peer
I230729 13:12:42.357065 89223 jobs/registry.go:1588  [T1,n1] 5505  AUTO SPAN CONFIG RECONCILIATION job 886590541109985281: stepping through state succeeded
E230729 13:12:42.357188 89223 jobs/registry.go:989  [T1,n1] 5506  error getting live session: node unavailable; try another peer
W230729 13:12:42.362003 99564 jobs/adopt.go:518  [T1,n1] 5507  could not clear job claim: clear-job-claim: failed to send RPC: sending to all replicas failed; last error: ba: Scan [/Table/15/1/886590545141465089,/Table/15/1/886590545141465090), [txn: bf2de98b], [can-forward-ts] RPC error: node unavailable; try another peer
W230729 13:12:42.362876 94965 jobs/adopt.go:518  [T1,n1] 5508  could not clear job claim: clear-job-claim: node unavailable; try another peer
I230729 13:12:42.366838 87603 server/server_controller_orchestration.go:268  [T1,n1] 5509  server controller shutting down ungracefully
I230729 13:12:42.366873 87603 server/server_controller_orchestration.go:279  [T1,n1] 5510  waiting for tenant servers to report stopped
W230729 13:12:42.366963 87603 server/server_sql.go:1704  [T1,n1] 5511  server shutdown without a prior graceful drain
--- FAIL: TestTxnObeysTableModificationTime (1.48s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #107736 sql/catalog/lease: TestTxnObeysTableModificationTime failed [C-test-failure O-robot T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnObeysTableModificationTime.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/catalog/lease/BUILD.bazel
+++ b/pkg/sql/catalog/lease/BUILD.bazel
@@ -77,6 +77,7 @@ go_test(
     embed = [":lease"],
     deps = [
         "//pkg/base",
+        "//pkg/jobs",
         "//pkg/jobs/jobspb",
         "//pkg/keys",
         "//pkg/kv",
@@ -105,7 +106,6 @@ go_test(
         "//pkg/sql/rowenc/keyside",
         "//pkg/sql/sem/tree",
         "//pkg/sql/sessiondata",
-        "//pkg/sql/sqltestutils",
         "//pkg/sql/tests",
         "//pkg/sql/types",
         "//pkg/storage",
@@ -114,7 +114,6 @@ go_test(
         "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
-        "//pkg/util",
         "//pkg/util/admission",
         "//pkg/util/encoding",
         "//pkg/util/hlc",

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -24,6 +24,7 @@ import (
 
 	"github.com/cockroachdb/cockroach-go/v2/crdb"
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv"
@@ -49,7 +50,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc/keyside"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
-	"github.com/cockroachdb/cockroach/pkg/sql/sqltestutils"
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
@@ -57,7 +57,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
-	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -981,219 +980,105 @@ func TestTxnObeysTableModificationTime(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	params := createTestServerParams()
-	params.ScanMaxIdleTime = time.Millisecond
-	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
+	s, sqlDB, _ := serverutils.StartServer(t, params)
 	defer s.Stopper().Stop(context.Background())
 
-	// Disable strict GC TTL enforcement because we're going to shove a zero-value
-	// TTL into the system with AddImmediateGCZoneConfig.
-	defer sqltestutils.DisableGCTTLStrictEnforcement(t, sqlDB)()
-
-	_, err := sqlDB.Exec(`SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '1s';`)
-	require.NoError(t, err)
-
-	// Refresh protected timestamp cache immediately to make MVCC GC queue to
-	// process GC immediately.
-	_, err = sqlDB.Exec(`SET CLUSTER SETTING kv.protectedts.poll_interval = '1s';`)
-	require.NoError(t, err)
-
-	// This test intentionally relies on uncontended transactions not being pushed
-	// in order to verify what it claims to verify. The default closed timestamp
-	// interval in 20.1+ is 3s. When run under the race detector, the process can
-	// stall for upwards of 3s leading to the write transaction getting pushed.
-	//
-	// In order to mitigate that push, we increase the target_duration when the
-	// test is run under race.
-	if util.RaceEnabled || skip.Stress() {
-		_, err := sqlDB.Exec(
-			"SET CLUSTER SETTING kv.closed_timestamp.target_duration = '120s'")
-		require.NoError(t, err)
-	}
-
-	if _, err := sqlDB.Exec(`
+	_, err := sqlDB.Exec(`
 CREATE DATABASE t;
 CREATE TABLE t.kv (k CHAR PRIMARY KEY, v CHAR);
 INSERT INTO t.kv VALUES ('a', 'b');
-`); err != nil {
-		t.Fatal(err)
+`)
+	require.NoError(t, err)
+
+	// requireOneRow ensures `res` contains only one row with two
+	// columns ("a", "b").
+	requireOneRow := func(res *gosql.Rows) {
+		for res.Next() {
+			var k, v, m string
+			require.Error(t, res.Scan(&k, &v, &m))
+			require.NoError(t, res.Scan(&k, &v))
+			require.Equal(t, "a", k)
+			require.Equal(t, "b", v)
+		}
+		require.NoError(t, res.Close())
 	}
-	tableDesc := desctestutils.TestingGetPublicTableDescriptor(kvDB, keys.SystemSQLCodec, "t", "kv")
+
+	// A helper to assert err is a transaction restart error with an error string
+	// that matches the supplied regex.
+	requireRestartTransactionErrWithMsg := func(t *testing.T, err error, re string) {
+		var pqe (*pq.Error)
+		if !errors.As(err, &pqe) || pgcode.MakeCode(string(pqe.Code)) != pgcode.SerializationFailure ||
+			!testutils.IsError(err, re) {
+			t.Fatalf("expected a %v error, got: %v", re, err)
+		}
+	}
+
+	// requireWriteTooOldErr ensures `err` is a WriteTooOldError.
+	requireWriteTooOldErr := func(t *testing.T, err error) {
+		requireRestartTransactionErrWithMsg(t, err, "WriteTooOldError")
+	}
+
+	// requireSessionExpiredErr ensures `err` is a liveness session expired error.
+	requireSessionExpiredErr := func(t *testing.T, err error) {
+		requireRestartTransactionErrWithMsg(t, err, "liveness session expired")
+	}
 
 	// A read-write transaction that uses the old version of the descriptor.
 	txReadWrite, err := sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
-
+	require.NoError(t, err)
 	// A read-only transaction that uses the old version of the descriptor.
 	txRead, err := sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
-
+	require.NoError(t, err)
 	// A write-only transaction that uses the old version of the descriptor.
 	txWrite, err := sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
 
 	// Modify the table descriptor.
-	if _, err := sqlDB.Exec(`ALTER TABLE t.kv ADD m CHAR DEFAULT 'z';`); err != nil {
-		t.Fatal(err)
-	}
+	_, err = sqlDB.Exec(`ALTER TABLE t.kv ADD m CHAR DEFAULT 'z';`)
+	require.NoError(t, err)
+	// Wait a short bit so that the SCHEMA CHANGE GC job created by the above ADD COLUMN
+	// gets to run, which will mark the old primary index of `t.kv` as tombstoned.
+	time.Sleep(1 * time.Second)
 
-	rows, err := txReadWrite.Query(`SELECT * FROM t.kv`)
-	if err != nil {
-		t.Fatal(err)
-	}
+	// Read-only transaction:
+	// 1). reads uses a historical version of `t.kv` and sees only two columns `k`
+	// and `v`;
+	// 2). it commits just fine;
+	rows, err := txRead.Query(`SELECT * FROM t.kv`)
+	require.NoError(t, err)
+	requireOneRow(rows)
+	require.NoError(t, txRead.Commit())
 
-	checkSelectResults := func(rows *gosql.Rows) {
-		defer func() {
-			if err := rows.Close(); err != nil {
-				t.Fatal(err)
-			}
-		}()
-		for rows.Next() {
-			// The transaction is unable to see column m.
-			var k, v, m string
-			if err := rows.Scan(&k, &v, &m); !testutils.IsError(
-				err, "expected 2 destination arguments in Scan, not 3",
-			) {
-				t.Fatalf("err = %v", err)
-			}
-			err = rows.Scan(&k, &v)
-			if err != nil {
-				t.Fatal(err)
-			}
-			if k != "a" || v != "b" {
-				t.Fatalf("didn't find expected row: %s %s", k, v)
-			}
-		}
-	}
+	// Read-write transaction:
+	// 1). reads uses a historical version of `t.kv` and sees only two columns `k`
+	// and `v`;
+	// 2). writes to that historical version of `t.kv` attempt to write to the old
+	// primary index, which has previously been written with a higher timestamp
+	// (when the SCHEMA CHANGE GC job lays a range tombstone on it). This is a
+	// write-write conflict, and it will bump up transaction's write_ts, perform a
+	// read refresh which would fail (bc reading from that new and higher
+	// timestamp would return zero row, instead of one), and thus a
+	// TransactionRetryWriteTooOld error will be returned.
+	rows, err = txReadWrite.Query(`SELECT * FROM t.kv`)
+	require.NoError(t, err)
+	requireOneRow(rows)
+	_, err = txReadWrite.Exec(`INSERT INTO t.kv VALUES ('c', 'd');`)
+	requireWriteTooOldErr(t, err)
+	require.NoError(t, txReadWrite.Rollback())
 
-	checkSelectResults(rows)
-
-	rows, err = txRead.Query(`SELECT * FROM t.kv`)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	checkSelectResults(rows)
-
-	// Read-only transaction commits just fine.
-	if err := txRead.Commit(); err != nil {
-		t.Fatal(err)
-	}
-
-	// This INSERT will cause the transaction to be pushed,
-	// which will be detected when we attempt to Commit() below.
-	if _, err := txReadWrite.Exec(`INSERT INTO t.kv VALUES ('c', 'd');`); err != nil {
-		t.Fatal(err)
-	}
-
-	checkDeadlineErr := func(err error, t *testing.T) {
-		var pqe (*pq.Error)
-		if !errors.As(err, &pqe) || pgcode.MakeCode(string(pqe.Code)) != pgcode.SerializationFailure ||
-			!testutils.IsError(err, "RETRY_COMMIT_DEADLINE_EXCEEDED") {
-			t.Fatalf("expected deadline exceeded, got: %v", err)
-		}
-	}
-
-	// The transaction read at one timestamp and wrote at another so it
-	// has to be restarted because the spans read were modified by the backfill.
-	checkDeadlineErr(txReadWrite.Commit(), t)
-
-	// This INSERT will cause the transaction to be pushed transparently,
-	// which will be detected when we attempt to Commit() below only because
-	// a deadline has been set.
-	if _, err := txWrite.Exec(`INSERT INTO t.kv VALUES ('c', 'd');`); err != nil {
-		t.Fatal(err)
-	}
-
-	checkDeadlineErr(txWrite.Commit(), t)
-
-	// Test the deadline exceeded error with a CREATE/DROP INDEX.
-	txWrite, err = sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
-	txUpdate, err := sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	// Modify the table descriptor.
-	if _, err := sqlDB.Exec(`CREATE INDEX foo ON t.kv (v)`); err != nil {
-		t.Fatal(err)
-	}
-
-	// This INSERT will cause the transaction to be pushed transparently,
-	// which will be detected when we attempt to Commit() below only because
-	// a deadline has been set.
-	if _, err := txWrite.Exec(`INSERT INTO t.kv VALUES ('c', 'd');`); err != nil {
-		t.Fatal(err)
-	}
-
-	checkDeadlineErr(txWrite.Commit(), t)
-
-	if _, err := txUpdate.Exec(`UPDATE t.kv SET v = 'c' WHERE k = 'a';`); err != nil {
-		t.Fatal(err)
-	}
-
-	checkDeadlineErr(txUpdate.Commit(), t)
-
-	txWrite, err = sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
-	txRead, err = sqlDB.Begin()
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	// Modify the table descriptor.
-	if _, err := sqlDB.Exec(`DROP INDEX t.kv@foo`); err != nil {
-		t.Fatal(err)
-	}
-
-	rows, err = txRead.Query(`SELECT k, v FROM t.kv@foo`)
-	if err != nil {
-		t.Fatal(err)
-	}
-	checkSelectResults(rows)
-
-	// Uses old descriptor and inserts values into index span which
-	// will be cleaned up.
-	if _, err := txWrite.Exec(`INSERT INTO t.kv VALUES ('c', 'd');`); err != nil {
-		t.Fatal(err)
-	}
-
-	if err := txRead.Commit(); err != nil {
-		t.Fatal(err)
-	}
-	if err := txWrite.Commit(); err != nil {
-		t.Fatal(err)
-	}
-
-	tableSpan := tableDesc.TableSpan(keys.SystemSQLCodec)
-
-	// Allow async schema change waiting for GC to complete (when dropping an
-	// index) and clear the index keys.
-	if _, err := sqltestutils.AddImmediateGCZoneConfig(sqlDB, tableDesc.GetID()); err != nil {
-		t.Fatal(err)
-	}
-
-	// Reset closed timestamp so that deleted keys can be GC'ed within the
-	// SucceedSoon window.
-	if util.RaceEnabled || skip.Stress() {
-		_, err := sqlDB.Exec(
-			"SET CLUSTER SETTING kv.closed_timestamp.target_duration = '3s'")
-		require.NoError(t, err)
-	}
-
-	testutils.SucceedsSoon(t, func() error {
-		return tests.CheckKeyCountIncludingTombstonedE(t, s, tableSpan, 2)
-	})
+	// Write-only transaction:
+	// 1). writes, similarly to the read-write transaction, will trigger a read
+	// refresh and this time it will succeed (bc there is no reads).
+	// 2). commits, however, will update the transaction's deadline, and the
+	// leased descriptor such an "old" transaction is using is the very original
+	// version (v1) of `t.kv` that expires at the modification time of v2 of
+	// `t.kv`. But the bumped commit timestamp is larger than the leased
+	// descriptor's expiration time, leading to a "liveness session expired"
+	// error.
+	_, err = txWrite.Exec(`INSERT INTO t.kv VALUES ('c', 'd');`)
+	require.NoError(t, err)
+	requireSessionExpiredErr(t, txWrite.Commit())
 }
 
 // Test that a lease on a table descriptor is always acquired on the latest

